WorkflowSwiftUI/Sources/ObservableScreen.swift

```diff
@@ -146,21 +147,31 @@ private final class ModeledHostingController<Model, Content: View>: UIHostingCon
 
         defer { hasLaidOutOnce = true }
 
-        if #available(iOS 16.0, *) {
-            // Handled in initializer, but set it on first layout to resolve a bug where the PCS is
-            // not updated appropriately after the first layout.
-            // UI-5797
-            if !hasLaidOutOnce,
-               swiftUIScreenSizingOptions.contains(.preferredContentSize)
-            {
-                let size = view.sizeThatFits(view.frame.size)
-
-                if preferredContentSize != size {
-                    preferredContentSize = size
-                }
-            }
-        } else if swiftUIScreenSizingOptions.contains(.preferredContentSize) {
-            let size = view.sizeThatFits(view.frame.size)
+        if swiftUIScreenSizingOptions.contains(.preferredContentSize) {
+            // Use the largest frame ever laid out in as a constraint for preferredContentSize
+            // measurements.
+            let width = max(view.frame.width, maxFrameWidth)
+            let height = max(view.frame.height, maxFrameHeight)
+
+            maxFrameWidth = width
+            maxFrameHeight = height
+
+            let fixedSize = CGSize(width: width, height: height)
+
+            // Measure a few different ways to account for ScrollView behavior. ScrollViews will
+            // always greedily fill the space available, but will report the natural content size
+            // when given an infinite size. By combining the results of these measurements we can
+            // deduce the natural size of content that scrolls in either direction, or both, or
+            // neither.
+
+            let fixedResult = view.sizeThatFits(fixedSize)
```

Comment:
Any idea what the perf cost is for these repeated measurements? I wonder if we could introduce our own sizing options that would allow us to do fewer calls based on more specific sizing configurations.

Reply:
also, is the new logic specifically for scroll views? if so, do we want to do some of these extra calculations conditionally? or does it have to account for them being embedded as descendant views?

Reply:
> Any idea what the perf cost is for these repeated measurements

Unknown, but my suspicion is it's fine due to SwiftUI layout being much better than what we're used to.

> also, is the new logic specifically for scroll views? if so, do we want to do some of these extra calculations conditionally? or does it have to account for them being embedded as descendant views?

We do need to calculate PCS regardless. The multiple measurements are to account for scroll views. Unfortunately we have to account for there being a scroll view anywhere in the content, and there's just no way to know.

Reply:
> Unfortunately we have to account for there being a scroll view anywhere in the content, and there's just no way to know.

It'd be a science experiment, but I wonder how feasible it'd be to push the measurements onto the SwiftUI side with a custom Layout and then bubble up context like 'hasScrollView'/'hasMultilineText' via preferences so we could be slightly smarter with our measurements.

I think I'd rather spend the time submitting feedback to Apple to hopefully make this better directly out of the box with the hosting controller though!